src/Components/Components/src/Auth/IAuthenticationStateProvider.cs

```diff
@@ -0,0 +1,33 @@
+// Copyright (c) .NET Foundation. All rights reserved.
+// Licensed under the Apache License, Version 2.0. See License.txt in the project root for license information.
+
+using System.Threading.Tasks;
+
+namespace Microsoft.AspNetCore.Components
+{
+    /// <summary>
+    /// Provides information about the authentication state of the current user.
+    /// </summary>
+    public interface IAuthenticationStateProvider
+    {
+        /// <summary>
+        /// Gets an <see cref="IAuthenticationState"/> instance that describes
+        /// the current user.
+        /// </summary>
+        /// <param name="forceRefresh">If true, instructs the provider to re-determine the user's authentication state, which can be an expensive operation. If false, the provider may reuse data cached for the current user.</param>
+        /// <returns>An <see cref="IAuthenticationState"/> instance that describes the current user.</returns>
+        Task<IAuthenticationState> GetAuthenticationStateAsync(bool forceRefresh);
```

Comment:
What would be a good scenario for ForceRefresh = true?

Reply:
When you're performed some action that you think may have invalidated the state (e.g., issued a logout request), then you need some way of checking how the state has now changed. This is only relevant in the client-side Blazor case (or custom `IAuthenticationStateProvider`) because that's the only place where it's even possible to change auth state without reloading the page.

The actual API may change. It looks like this in the spec but it could end up being a separate method (`RefreshAuthenticationState()`) by the time we review the end product.

Reply:
I think that is implicit in your `IAuthenticationStateProvider.cs` implementation. Your IAuthenticationStateProvider should be responsible for determining with absolute certainty whether you are authenticated or not and hide that from the consumers.

I think the better model is for components to be reactive and to subscribe to the event that IAuthenticationStateProvider give you to get notified of changes. Whatever the implementation is for IAuthenticationStateProvider, that component should be the single source of truth for login and logout operations.

In the case you are using cookies, the cookie expiration determines how long you will be authenticated. (And there's no out of the box way for us to track that other than setting up some sort of timer or checking the cookie expiration on every call (Which we should do anyways as we don't want to provide a stale value when we know its no longer valid))

In the case you are using oAuth/JWTs, the token can be invalidated between when you check and when the actual API call/Rendered view happens, so you need to plan for that case anyways.

What I'm trying to convene here, is that components/services should react to the changes on the auth state and plan for the value being possibly stale rather than proactively checking against a source to see that things are up to date.

Reply:
> Your IAuthenticationStateProvider should be responsible for determining with absolute certainty whether you are authenticated or not and hide that from the consumers.

I disagree for the client-side case. The client isn't the source of truth (the server is). We can't re-query the server on every call to determine this, and even if we did, the state might already have changed again since. On the client, the `IAuthenticationStateProvider` is tracking what the server has most recently told us.

> I think the better model is for components to be reactive and to subscribe to the event that IAuthenticationStateProvider give you to get notified of changes.

That already does happen.

> What I'm trying to convene here, is that components/services should react to the changes on the auth state and plan for the value being possibly stale rather than proactively checking against a source to see that things are up to date.

Perhaps I was unclear in my description. It's not expected that typical UI components are responsible for proactively querying for updated authentication state. Typical UI components just receive the cascading parameter and react to it.

The only time the developer would be concerned about invalidating auth state is:

 * They are running client-side, or have a custom `IAuthenticationStateProvider`
 * They have taken some action that, due to application-specific rules that only they can know about, may have changed something about the `ClaimsPrincipal`. For example maybe the user just subscribed to "gold tier" service and now may have a new claim value, but only the server really knows for sure. The UI would like to get refreshed info, in case doing so causes some other components to display new output (e.g., different menu entries).

Does this make sense?

Reply:
To help me track remaining todos, I'll mark this as resolved since it seems like the conversation concluded.